test/Ocelot.UnitTests/Kubernetes/KubeTests.cs

```diff
@@ -54,6 +56,66 @@ public async Task Should_return_service_from_k8s()
         receivedToken.Value.ShouldBe($"Bearer {nameof(Should_return_service_from_k8s)}");
     }
 
+    [Theory]
+    [InlineData(HttpStatusCode.BadRequest)]
+    [InlineData(HttpStatusCode.Forbidden)]
+    [InlineData(HttpStatusCode.InternalServerError)]
+    [InlineData(HttpStatusCode.NotFound)]
+    [Trait("Feat", "345")]
+    public async Task Should_not_return_service_from_k8s_when_k8s_api_returns_error_response(HttpStatusCode expectedStatusCode)
+    {
+        // Arrange
+        var given = GivenClientAndProvider(out var serviceBuilder);
+        serviceBuilder.Setup(x => x.BuildServices(It.IsAny<KubeRegistryConfiguration>(), It.IsAny<EndpointsV1>()))
+            .Returns(new Service[] { new(nameof(Should_return_service_from_k8s), new("localhost", 80), string.Empty, string.Empty, Array.Empty<string>()) });
+
+        var endpoints = GivenEndpoints();
+        using var kubernetes = GivenThereIsAFakeKubeServiceDiscoveryProvider(
+            given.ClientOptions.ApiEndPoint.ToString(),
+            given.ProviderOptions.KubeNamespace,
+            given.ProviderOptions.KeyOfServiceInK8s,
+            expectedStatusCode,
+            endpoints,
+            out Lazy<string> receivedToken);
+
+        string expectedKubeApiErrorMessage = GetKubeApiErrorMessage(serviceName: given.ProviderOptions.KeyOfServiceInK8s, given.ProviderOptions.KubeNamespace, expectedStatusCode);
+        string expectedLogMessage = $"Failed to retrieve v1/Endpoints '{given.ProviderOptions.KeyOfServiceInK8s}' in namespace '{given.ProviderOptions.KubeNamespace}' (HTTP.{expectedStatusCode}/Failure/{expectedStatusCode}): {expectedKubeApiErrorMessage}";
+        _logger.Setup(logger => logger.LogError(It.IsAny<Func<string>>(), It.IsAny<Exception>()))
+            .Callback((Func<string> messageFactory, Exception exception) =>
+            {
+                messageFactory.ShouldNotBeNull();
+
+                string logMessage = messageFactory();
+                logMessage.ShouldNotBeNullOrWhiteSpace();
+
+                // This is a little fragile, as it may change if other entries are logged due to implementation changes.
```

Comment:
@raman-m small issue, as noted in these comments.

Reply:
Wow!... 😮

Reply:
The problem (as it looks to me) is that the test can't know for sure what other components may log messages as well (in this case, the retry facility uses the same logger and trying to verify those logged messages would make the whole thing too fragile by making this unit test coupled to the implementation of other components such as the retry).

I've always aimed to make unit tests, in particular, to test behaviour without being coupled to the implementation; otherwise the tests may have to be changed each time the implementation changes which removes one of the (in my opinion) best benefits of unit testing.

But I'm not that familiar with the code-base and so I'm open to changing it if you believe that would be better 🙂

Reply:
So, the issue lies in the retry logic of the upper context, correct? It seems we have been debating this extensively. If the problem truly exists, I will identify it in the next round of refactoring the logic. For now, we can omit this.